File: 30DaysOfCode/day28_regExPatternsAndIntroToDatabases/regExPatternsAndIntroToDatabases.js
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    const N = parseInt(readLine(), 10);
    const pattern = new RegExp(".+@gmail\\.com$");
    const names = [];
    
    for (let NItr = 0; NItr < N; NItr++) {
        const firstNameEmailID = readLine().split(' ');
        const firstName = firstNameEmailID[0];
        const emailID = firstNameEmailID[1];
        
        if (pattern.test(emailID)) {
            names.push(firstName);
        }
    }
    
    names.sort();
    for (let i = 0; i < names.length; i++) {
        console.log(names[i]);
    }
}

File: 30DaysOfCode/day28_regExPatternsAndIntroToDatabases/regExPatternsAndIntroToDatabases.pl
#!/usr/bin/perl

use strict;
use warnings;

my %names = ();

my $N = <>;
$N =~ s/\s+$//;

for (my $N_itr = 0; $N_itr < $N; $N_itr++) {
    my $firstNameEmailID = <>;
    $firstNameEmailID =~ s/\s+$//;
    my @firstNameEmailID = split /\s+/, $firstNameEmailID;

    my $firstName = $firstNameEmailID[0];
    chomp($firstName);

    my $emailID = $firstNameEmailID[1];
    chomp($emailID);
    
    if ($emailID =~ /\w+\@gmail\.com$/) {
        $names{$emailID} = $firstName;
    }
}

foreach $i (keys %names) {
    push @names, $names{$i};
}

print join "\n", sort @names;
